File: archive/previous_campaign/tools/ghidra/scripts/export_decomp.py
# Ghidra headless script: export decompiled C for a list of function addresses
#
# Invocation (via analyzeHeadless):
#   -postScript export_decomp.py <path/to/function_list.txt> <output_dir>
#
# function_list.txt format: one hex address per line (with or without 0x prefix)
# Lines starting with # are comments.
#
# Output: <output_dir>/fn_<ADDR>_ghidra.c with the Ghidra decompiler's C for
# that function. Addresses that don't resolve to a function are logged.
#@category Pokemon

import os

from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor


def parse_list(path):
    addrs = []
    with open(path, "r") as f:
        for raw in f:
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            line = line.split("#")[0].strip()
            if line.lower().startswith("0x"):
                line = line[2:]
            try:
                addrs.append(int(line, 16))
            except ValueError:
                print("skipping unparseable line: %r" % raw)
    return addrs


def run(program, list_path, out_dir):
    if not os.path.isdir(out_dir):
        os.makedirs(out_dir)

    decomp = DecompInterface()
    opts = DecompileOptions()
    decomp.setOptions(opts)
    decomp.openProgram(program)

    monitor = ConsoleTaskMonitor()
    fn_mgr = program.getFunctionManager()
    space = program.getAddressFactory().getDefaultAddressSpace()

    addrs = parse_list(list_path)
    print("exporting %d functions" % len(addrs))
    ok = 0
    fail = 0

    for a in addrs:
        addr = space.getAddress(a)
        fn = fn_mgr.getFunctionAt(addr)
        if fn is None:
            fn = fn_mgr.getFunctionContaining(addr)
        if fn is None:
            print("no function at 0x%08x" % a)
            fail += 1
            continue

        result = decomp.decompileFunction(fn, 120, monitor)
        if result is None or not result.decompileCompleted():
            msg = result.getErrorMessage() if result else "no result"
            print("decompile failed for fn_%08X: %s" % (a, msg))
            fail += 1
            continue

        code = result.getDecompiledFunction().getC()
        out_path = os.path.join(out_dir, "fn_%08X_ghidra.c" % a)
        with open(out_path, "w") as f:
            f.write(code)
        ok += 1

    print("exported %d, failed %d" % (ok, fail))


args = getScriptArgs()
if len(args) < 2:
    print("usage: export_decomp.py <function_list.txt> <output_dir>")
else:
    run(currentProgram, args[0], args[1])

File: archive/previous_campaign/tools/ghidra/scripts/ExportDecomp.java
// Ghidra headless script: export decompiled C for a list of function addresses.
// Invocation via analyzeHeadless:
//   -postScript ExportDecomp.java <path/to/function_list.txt> <output_dir>
// @category Pokemon

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.List;

public class ExportDecomp extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 2) {
            println("usage: ExportDecomp <function_list.txt> <output_dir>");
            return;
        }
        String listPath = args[0];
        String outDir = args[1];

        File outDirFile = new File(outDir);
        if (!outDirFile.exists()) outDirFile.mkdirs();

        List<Long> addrs = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(listPath))) {
            String line;
            while ((line = br.readLine()) != null) {
                String s = line.trim();
                if (s.isEmpty() || s.startsWith("#")) continue;
                int hash = s.indexOf('#');
                if (hash >= 0) s = s.substring(0, hash).trim();
                if (s.startsWith("0x") || s.startsWith("0X")) s = s.substring(2);
                try {
                    addrs.add(Long.parseLong(s, 16));
                } catch (NumberFormatException e) {
                    println("skipping unparseable: " + line);
                }
            }
        }
        println("exporting " + addrs.size() + " functions");

        DecompInterface decomp = new DecompInterface();
        decomp.setOptions(new DecompileOptions());
        decomp.openProgram(currentProgram);

        ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();
        FunctionManager fnMgr = currentProgram.getFunctionManager();
        int ok = 0;
        int fail = 0;

        for (Long a : addrs) {
            Address addr = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(a);
            Function fn = fnMgr.getFunctionAt(addr);
            if (fn == null) fn = fnMgr.getFunctionContaining(addr);
            if (fn == null) {
                println(String.format("no function at 0x%08x", a));
                fail++;
                continue;
            }
            DecompileResults result = decomp.decompileFunction(fn, 120, monitor);
            if (result == null || !result.decompileCompleted()) {
                String msg = (result != null) ? result.getErrorMessage() : "null";
                println(String.format("decompile failed for fn_%08X: %s", a, msg));
                fail++;
                continue;
            }
            String code = result.getDecompiledFunction().getC();
            String outPath = new File(outDirFile, String.format("fn_%08X_ghidra.c", a)).getPath();
            try (FileWriter w = new FileWriter(outPath)) {
                w.write(code);
            }
            ok++;
        }
        println("exported " + ok + ", failed " + fail);
        decomp.dispose();
    }
}
